### PR TITLE
HW-01: Bubble sort visualisation

### PR DESCRIPTION
**Pull request by vbarbasins**
# Hometask

## Changes
- Added simple app with bubble sort visualisation
- Bar height in pixels represent number values, bar with current sorting state highlighted 
- Constants for number count, max number values and update timeout can be adjusted in code
- Can be started, paused, created new number set
- When sorting finished - start button disables

## Demo

Project build published to http://vb-bubble-sort.surge.sh

https://user-images.githubusercontent.com/23147889/114277279-3483e200-9a33-11eb-91a3-bd7f8f2c86c2.mp4

### CODE REVIEW

File: homeworks/01-bubble-sort/src/App.tsx [hunk 1]
```diff
@@ -1,26 +1,103 @@
...
+  sorting: boolean,
+  sortingState: {
+    iteration: number,
+    currentIndex: number,
+  },
+  numbers: number[],
+  solved: boolean,
+}
+
+class App extends React.Component<{}, AppState> {
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx:22]: You have `AppProps `, but don't use it here.

File: homeworks/01-bubble-sort/src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,103 @@
...
+  constructor(props: AppProps) {
+    super(props);
+    this.state = {
+      sorting: false,
+      sortingState: { iteration: 0, currentIndex: 0 },
+      numbers: createRandomNumberArray(MAX_NUMBER, NUMBER_COUNT),
+      solved: false,
+    };
+  }
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx]: You can write just the following:
```tsx
    state = {
      sorting: false,
      sortingState: { iteration: 0, currentIndex: 0 },
      numbers: createRandomNumberArray(MAX_NUMBER, NUMBER_COUNT),
      solved: false,
    };
```

File: homeworks/01-bubble-sort/src/App.tsx [hunk 3]
```diff
@@ -1,26 +1,103 @@
...
+      sorting: false,
+      sortingState: { iteration: 0, currentIndex: 0 },
+      numbers: createRandomNumberArray(MAX_NUMBER, NUMBER_COUNT),
+      solved: false,
+    });
+  }
+
+  progressHandler = (): void => {
+    const { sorting } = this.state;
+    this.setState((prevState) => ({ ...prevState, sorting: !sorting }));
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx]: You don't need `...prevState`, `setState` merges objects by default.
```tsx
const { sorting } = this.state;
this.setState({ sorting: !sorting });
```
or
```tsx
this.setState(({ sorting }) => ({ sorting: !sorting }));
```
Both work the same.

File: homeworks/01-bubble-sort/src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,103 @@
...
+  }
+
+  sortNumbers = (): void => {
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx]: Better to keep a business logic separate from the component. E.g., create a pure function and move it to another file (e.g., under `utils` folder).

File: homeworks/01-bubble-sort/src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,103 @@
...
+            currentIndex: continueIteration ? nextIndex : 0,
+          },
+        };
+      });
+    } else {
+      this.setState((prevState) => ({
+        ...prevState,
+        sorting: false,
+        solved: true,
+      }));
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx]: Same:
```tsx
this.setState({
  sorting: false,
  solved: true,
});
```

File: homeworks/01-bubble-sort/src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,103 @@
...
+      const nextIndex = currentIndex + 1;
+      if (numbers[currentIndex] > numbers[nextIndex]) {
+        const tmp = newNumbers[currentIndex];
+        newNumbers[currentIndex] = newNumbers[nextIndex];
+        newNumbers[nextIndex] = tmp;
+      }
+      this.setState((prevState) => {
+        const continueIteration = nextIndex < numbers.length - iteration;
+        return {
+          ...prevState,
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx]: Same

File: homeworks/01-bubble-sort/src/utils/appUtils.test.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { createRandomNumberArray } from './appUtils';
+
+describe('appUtils', () => {
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/utils/appUtils.test.ts:3]: Nice to see tests 👍

But they would be more applicable to `sort` functionality.

**Review by shpakkdv — APPROVED** [review]
Overall, the code and the app look good.

Also, you have a bug: try to click `Start` and during sorting click `Reset`. Pay attention to the highlighted `current` value.

File: homeworks/01-bubble-sort/src/App.tsx (hunk 2 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: I get an error, if not using constructor:
> State initialization should be in a constructor  react/state-in-constructor

File: homeworks/01-bubble-sort/src/App.tsx (hunk 1 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx:22]: Fixed

File: homeworks/01-bubble-sort/src/App.tsx (hunk 3 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: fixed, thx! understood

File: homeworks/01-bubble-sort/src/App.tsx (hunk 5 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: thx, updated

File: homeworks/01-bubble-sort/src/App.tsx (hunk 6 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: Fixed.

File: homeworks/01-bubble-sort/src/utils/appUtils.test.ts (hunk 7 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/utils/appUtils.test.ts:3]: Yes, i will open an issue to implement unit tests for sort functionality. Will implement later.

Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/utils/appUtils.test.ts:3]: https://github.com/vbarbasins/evo-ts-bootcamp/issues/5

File: homeworks/01-bubble-sort/src/App.tsx (hunk 4 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: Please, check https://github.com/vbarbasins/evo-ts-bootcamp/pull/1/commits/045cda6392d132bcdfc199f292704b6fe6b12103

File: homeworks/01-bubble-sort/src/utils/appUtils.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        currentIndex: continueIteration ? nextIndex : 0,
+      },
+    };
+  }
+  return {
+    ...state,
+    sorting: false,
+    solved: true,
+  };
+};
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/utils/appUtils.ts]: Utils shouldn't know anything about components, state, etc. They should be just pure functions if possible. No needs to use components interfaces.

File: homeworks/01-bubble-sort/src/App.tsx [hunk 9]
```diff
@@ -1,26 +1,78 @@
...
+      solved: false,
+    };
+  }
+
+  componentDidUpdate(): void {
+    const { sorting } = this.state;
+    if (sorting) setTimeout(this.sortNumbers, SORT_TIMEOUT);
+  }
+
+  newSetHandler = (): void => {
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx]: Always define access modifiers.

File: homeworks/01-bubble-sort/src/App.tsx (hunk 2 above)
Comment (shpakkdv) [reply homeworks/01-bubble-sort/src/App.tsx]: This is airbnb preference, but both options are correct 🙂

File: homeworks/01-bubble-sort/src/components/BarSet.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+export const BarSet: React.FC<BarSetProps> = ({ numbers, currentIndex }) => (
+  <div className="BarSet">
+    {numbers.map((number, index) => {
+      const style = {
+        height: `${number}px`,
+        backgroundColor: index === currentIndex ? '#fddeb6' : '#faebd7',
+      };
+      return <div className="Bar" style={style} key={`${index}`} />;
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/components/BarSet.tsx]: Just `key={index}`.

File: homeworks/01-bubble-sort/src/utils/appUtils.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  maxNumber: number,
+  numberCount: number,
+): number[] => {
+  const array: number[] = [];
+  while (array.length < numberCount) {
+    array.push(
+      Math.floor(Math.random() * maxNumber) + 1,
+    );
+  }
+  return array;
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/utils/appUtils.ts]: Should work the same:
`Array.from({ length: numberCount }, () => Math.floor(Math.random() * maxNumber) + 1)`

**Review by shpakkdv — APPROVED** [review]
The descried bug is still reproducible:

Also, you have a bug: try to click `Start`, and during sorting click `New set`. Pay attention to the highlighted `current` value.

File: homeworks/01-bubble-sort/src/utils/appUtils.ts (hunk 11 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/utils/appUtils.ts]: wow, nice, thx! Updated.

File: homeworks/01-bubble-sort/src/components/BarSet.tsx (hunk 10 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/components/BarSet.tsx]: fixed

File: homeworks/01-bubble-sort/src/App.tsx (hunk 2 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: disabled this rule, initialised state in the component beginning, remove useless AppProps interface

File: homeworks/01-bubble-sort/src/utils/appUtils.ts (hunk 8 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/utils/appUtils.ts]: Remade, please check commit https://github.com/vbarbasins/evo-ts-bootcamp/commit/947ad694cc25c800a62a0009bc5508b4984a33d5

Also bug fixed by checking if sorting is enabled before making sorting step.

File: homeworks/01-bubble-sort/src/App.tsx (hunk 9 above)
Comment (vbarbasins) [reply homeworks/01-bubble-sort/src/App.tsx]: added private here and other custom functions in component.

**Pull request merged**